oauth2_http/java/com/google/auth/oauth2/TokenVerifier.java

```diff
@@ -0,0 +1,404 @@
+  }
+
+  /**
+   * Verify an encoded JWT token.
+   *
+   * @param token encoded JWT token
+   * @return the parsed JsonWebSignature instance for additional validation if necessary
+   * @throws VerificationException thrown if any verification fails
+   */
+  public JsonWebSignature verify(String token) throws VerificationException {
+    JsonWebSignature jsonWebSignature;
+    try {
+      jsonWebSignature = JsonWebSignature.parse(OAuth2Utils.JSON_FACTORY, token);
+    } catch (IOException e) {
+      throw new VerificationException("Error parsing JsonWebSignature token", e);
+    }
+
+    // Verify the expected audience if an audience is provided in the verifyOptions
+    if (audience != null && !audience.equals(jsonWebSignature.getPayload().getAudience())) {
+      throw new VerificationException("Expected audience does not match");
+    }
+
+    // Verify the expected issuer if an issuer is provided in the verifyOptions
+    if (issuer != null && !issuer.equals(jsonWebSignature.getPayload().getIssuer())) {
+      throw new VerificationException("Expected issuer does not match");
+    }
+
+    Long expiresAt = jsonWebSignature.getPayload().getExpirationTimeSeconds();
+    if (expiresAt != null && expiresAt <= clock.currentTimeMillis() / 1000) {
+      throw new VerificationException("Token is expired");
+    }
+
+    // Short-circuit signature types
+    if (!SUPPORTED_ALGORITHMS.contains(jsonWebSignature.getHeader().getAlgorithm())) {
+      throw new VerificationException(
+          "Unexpected signing algorithm: expected either RS256 or ES256");
+    }
+
+    PublicKey publicKeyToUse = publicKey;
+    if (publicKeyToUse == null) {
+      try {
+        String certificateLocation = getCertificateLocation(jsonWebSignature);
+        publicKeyToUse =
+            publicKeyCache.get(certificateLocation).get(jsonWebSignature.getHeader().getKeyId());
+      } catch (ExecutionException | UncheckedExecutionException e) {
+        throw new VerificationException("Error fetching PublicKey from certificate location", e);
+      }
+    }
+
+    if (publicKeyToUse == null) {
+      throw new VerificationException(
+          "Could not find PublicKey for provided keyId: "
+              + jsonWebSignature.getHeader().getKeyId());
+    }
+
+    try {
+      if (jsonWebSignature.verifySignature(publicKeyToUse)) {
+        return jsonWebSignature;
+      }
+      throw new VerificationException("Invalid signature");
+    } catch (GeneralSecurityException e) {
+      throw new VerificationException("Error validating token", e);
+    }
+  }
+
+  private String getCertificateLocation(JsonWebSignature jsonWebSignature)
+      throws VerificationException {
+    if (certificatesLocation != null) return certificatesLocation;
+
+    switch (jsonWebSignature.getHeader().getAlgorithm()) {
+      case "RS256":
+        return FEDERATED_SIGNON_CERT_URL;
+      case "ES256":
+        return IAP_CERT_URL;
```

Comment:
I'm a little nervous about using the algorithm to determine the cert location, since feasibly there could be a new cert for `RS256` or `ES256` down the road. Would it be better to make this explicit, e.g. make `certificatesLocation` a required parameter? Or add methods for each cert location such as `verifyOIDC` and `verifyIAP` ?

Reply:
At least for verifying the signature, I don't think the developer would care to distinguish between an OICD token or IAP token.

If Google provides additional JWKS locations for different algorithms, I would expect us to change this behavior to look for all the default certification locations. I think we could accomplish that without changing the public interface.